R/UploadControls.R

```diff
@@ -0,0 +1,185 @@
+		  group_by(density,strain,percentage,batch) %>%
+		  dplyr::mutate(base.count=sum(!is.na(study_subject_id))) %>%
+		  dplyr::mutate(count = as.integer(count), count = sum(base.count + count)) %>%
+		  distinct() %>%
+		  dplyr::mutate(count = paste(1:count, collapse=";")) %>%
+			dplyr::mutate(count = strsplit(count, ";")) %>%
+			tidyr::unnest(cols=c("count")) %>%
+		  ungroup() %>%
+			dplyr::mutate(count=as.character(count)) %>%
+			tidyr::unite(control, c(density,strain,percentage,count), remove = FALSE) %>%
+		  mutate(created = now) %>%
+		  mutate(last_updated = now) %>%
+	    select(created, last_updated, control, study_id, all_of(colnames(user_data)))
+
+		df.payload$sheet_id = user_data %>%
+		  dplyr::mutate(count=as.integer(count)) %>%
+		  dplyr::mutate(sheet_id=row_number()) %>%
+		  tidyr::uncount(count) %>%
+		  pull(sheet_id)
+
+		## Add the new study subjects here
+		res <- dbAppendTable(con, "study_subject", df.payload %>% select(created, last_updated, study_id, control) %>% dplyr::rename(name = control))
+
+		## Rejoin to get the study_subject_id
+		df.payload = df.payload %>%
+			inner_join(dbReadTable(con, "study_subject") %>% dplyr::rename(study_subject_id=id), by = c("control"="name", "created", "last_updated", "study_id")) %>%
+		  select(created,last_updated,control,study_subject_id,sheet_id,all_of(colnames(user_data)))
+
+		## Find the locations and bags and join
+		df.payload = df.payload %>%
+		  inner_join(dbReadTable(con, "location") %>%
+		               select(-c(created, last_updated)) %>%
+		               dplyr::rename(location_id=id, location_name=name)
+		             , by = c("location_name", "level_I", "level_II")) %>%
+		  dplyr::left_join(dbReadTable(con, "dbs_bag") %>%
+		                     select(-c(created, last_updated)) %>%
+		                     dplyr::rename(dbs_bag_id=id), by=c("manifest_name"="name", "location_id")) %>%
+		  filter(is.na(dbs_bag_id)) %>%
+		  select(created,last_updated,control,study_subject_id,location_id,dbs_bag_id,sheet_id, all_of(colnames(user_data)))
+
+		## if the bag does not exist (dbs_bag_id == `NA`), then create it
+		res <- dbAppendTable(con, "dbs_bag", df.payload %>%
+		                       filter(is.na(dbs_bag_id)) %>%
+		                       select(created, last_updated, location_id, manifest_name) %>%
+		                       dplyr::rename(name = manifest_name) %>%
+		                       distinct())
+
+		df.payload=dbReadTable(con, "dbs_bag") %>% dplyr::rename(bag_id=id) %>%
+		  inner_join(df.payload %>% select(-c(created,last_updated)), by = c("location_id", "name"="manifest_name")) %>%
+		  dplyr::rename(manifest_name=name) %>%
+		  select(created,last_updated,control, study_subject_id,location_id,bag_id,sheet_id,all_of(colnames(user_data)))
+
+		## Multiple sheets can go into a bag. Instead of identifying them uniquely, add a unique integer
```

Comment:
The sheets need to be numbered/labeled in some way I think that is added to the database. Otherwise when you punch, how will you know which sheet to remove the count from?